Added Python migration guide

smythp opened the pull request:
## Type of change

Added Python migration guide to Migration section

### How should this PR be tested?

Testable code shared in the tutorial is provided in the [linked repository](https://github.com/chainguard-dev/cg-images-python-migration). Feel free to open issues in that repo if there are problems with the build, or comment here as part of the review.

smythp committed:
Added Python migration guide

Signed-off-by: Patrick Smyth <patrick.smyth@chainguard.dev>

amouat reviewed (approved):
Looks good.

I think it's probably worth having a "differences to the Docker Official Image" section is only big thing.

smythp committed:
Parallel structure, one typo

Signed-off-by: Patrick Smyth <patrick.smyth@chainguard.dev>

smythp commented on content/chainguard/migration/migrating-python.md:
```diff
@@ -0,0 +1,124 @@
+
+## Chainguard Images for Python Overview
+
+We distribute two versions of our [Python Chainguard Image](https://edu.chainguard.dev/chainguard/chainguard-images/reference/python/), one for development that includes shells such as bash and package managers such as pip and a production image that removes these tools for increased security. Our production images are tagged as `latest`, while our development images are tagged as `latest-dev`. 
```
Good distinction. I added the word "public" to make this more accurate. Wrote ash/bash and used sh, ash, and bash as examples in another spot.

smythp commented on content/chainguard/migration/migrating-python.md:
```diff
@@ -0,0 +1,124 @@
+
+As you can see, the primary difference in this Flask application compared to the pre-migration application is the use of a multistage build. In the initial stage, we copy our requirements into the development version of the Python Chainguard Image, initialize a virtual environment, and install needed packages with pip. In the second stage, we copy the virtual environment from the development imag ...
+
+By default, the entrypoint for the Python Chainguard Image is `python` rather than `bash`. However, if you replace the included Python binary with the virtual environment binary on the path as we do above, you should set the entrypoint explicitly, otherwise you will not have access to the packages included in your virtual environment.
```
No, they're symlinks, but using that word would confuse things I'll try to split the difference here and not use the word binary. They're executables, but we don't usually say that in Linux. I'll just use more vague terms like "the `python`"

smythp commented on content/chainguard/migration/migrating-python.md:
```diff
@@ -0,0 +1,124 @@
+
+We recommend that you pin dependencies to specific versions in your own application. The example Flask application script linked above also enables debug mode, which should be turned off in a production scenario.
+
+You may wish to include the following environmental variables in your Dockerfile. They prevent the buffering of output and prevent the creation of cached bytecode, respectively, and can be included in both stages of the multi-stage build.
```
I think these have an extremely marginal benefit, but they're actively recommended so I don't want to be remiss. This is the new language:

You may wish to include the following environmental variables in your Dockerfile. The first prevents the buffering of output, meaning that all messages printed to standard output are immediately printed rather than being held in a cache. The second prevents the creation of cached bytecode, which can marginally reduce image size.

smythp commented on content/chainguard/migration/migrating-python.md:
```diff
@@ -0,0 +1,124 @@
+
+## Serving an Application with nginx and Docker Compose
+
+We provide an [nginx Chainguard Image](/chainguard/chainguard-images/reference/nginx/), also with low to no CVEs, that can be used as a secure and performant reverse proxy to serve your application. You can view an [example orchestration of a Flask application and nginx using Chainguard Images](https://github.com/chainguard-dev/cg-images-python-migration/tree/compose-flask-nginx) at the linked re ...
```
It's not an adjective, so I don't know why it would have hyphens. 🤷 I could be persuaded, but I'll leave as-is for now.

This is how I would do it:

> It's a low-to-no-CVE image, so it's pretty cool.

> The image is pretty cool because it has low to no CVEs.

If it's a style guide thing then I'll change it. :)

smythp commented on content/chainguard/migration/migrating-python.md:
```diff
@@ -0,0 +1,124 @@
+    build: nginx
+    restart: always
+    ports:
+      - 80:80
```
Yeah I thought about that, but on consideration and testing it seems to be OK. I could use a higher port inside and 80 on the outside but I think I'll go with ain't broke on this one. If it wasn't going to work it'd not work inside the container. On the outside you pretty much have to serve on 80, in a real setup you'd sere on 80 and checkin the incoming domain and route. Had the same spidey sense, though.

smythp committed:
Updates following Adrian's review

Added section comparing official Docker Python and CG Image

Many smaller revisions

Signed-off-by: Patrick Smyth <patrick.smyth@chainguard.dev>

smythp commented:
Thanks @amouat for the detailed review . Added a pretty comprehensive comparison section after looking at your node guide
